### PR TITLE
strategy: add logic for weekly maintenance windows

### PR DESCRIPTION
**Pull request by lucab**
This implements a new update strategy, with support for weekly maintenance windows.

Closes: https://github.com/coreos/zincati/issues/34

### CODE REVIEW

**Comment by lucab** [comment]
Status update: rebased to master, lower level logic for weekly intervals should be complete now.

**Comment by lucab** [comment]
@jlebon this should be ready for an initial review pass, possibly starting top-down from the usage docs. I still plan to perform a couple more manual tests to make sure I didn't miss some corner cases.

File: src/config/inputs.rs [hunk 1]
```diff
@@ -176,6 +178,21 @@ pub(crate) struct FleetLockInput {
     pub(crate) base_url: String,
 }
 
+/// Config for "weekly" strategy.
```
Comment (jlebon) [inline src/config/inputs.rs:181]: Sidenote: I find the naming for this strategy a little confusing. "weekly" makes it seem like all it can do is do updates every week. When really, it also implements daily or every X hours, etc, right? I don't really have a better suggestion though... maybe "windowed"?

File: src/strategy/weekly.rs [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+            let start = utils::time_from_string(&entry.start_time)?;
+            let length = Duration::from_secs(u64::from(entry.length_minutes).saturating_mul(60));
+            let windows = DailyWindow::parse_timespan(weekday, start.0, start.1, length)?;
+            intervals.extend(windows);
+        }
+
+        let calendar = WeeklyCalendar::new(intervals);
+        match calendar.total_length_minutes() {
+            0 => failure::bail!("zero-length weekly updates window"),
+            n => log::trace!("weekly updates, total length: {} minutes", n),
```
Comment (jlebon) [inline src/strategy/weekly.rs:36]: Hmm, wouldn't it be more useful to report the non-overlapping total number of minutes here?

File: docs/usage/updates-strategy.md [hunk 3]
```diff
@@ -56,3 +56,40 @@ Such an approach is only recommended where nodes are already grouped into an orc
...
+start_time = "01:00"
+length_minutes = 30
+```
+
+The above configuration would result in three maintenance windows during which Zincati is allowed to reboot the machine for updates:
+ * 60 minutes starting at 23:30 UTC on Saturday night, and ending at 00:30 UTC on Sunday morning
+ * 60 minutes starting at 23:30 UTC on Sunday night, and ending at 00:30 UTC on Monday morning
+ * 30 minutes starting at 01:00 UTC on Wednesday morning, and ending at 01:30 UTC on Wednesday morning
+
+Reboot windows can be separetely configured in multiple snippets, as long as each `updates.weekly.window` entry contains all the required properties:
```
Comment (jlebon) [inline docs/usage/updates-strategy.md:90]: separately

File: src/strategy/weekly.rs (hunk 2 above)
Comment (lucab) [reply src/strategy/weekly.rs:36]: Absolutely yes, but I'd need to implement that in the underlying library first (I have it on my radar). This is a trace-log not seen by default and mostly meant for debugging.
Once we have the correct number, it should be better exposed via metrics or such.

File: src/config/inputs.rs (hunk 1 above)
Comment (lucab) [reply src/config/inputs.rs:181]: I do not disagree :)

In the first iteration here I used "periodic", but I realized it was too generic (and I'm still considering if I can have another strategy based on https://www.freedesktop.org/software/systemd/man/systemd.time.html# later on).

Locksmith used to just call this "reboot_window" (but it allowed a single one).

If we can find a single word which conveys both the "timespans on a calendar" and "recurring every week" I'd be very happy.

File: src/weekly/mod.rs [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+}
+
+impl DailyWindow {
+    /// Parse a timespan into daily windows.
+    pub fn parse_timespan(
+        start_day: chrono::Weekday,
+        start_hour: u8,
+        start_minute: u8,
+        length: Duration,
+    ) -> Fallible<Vec<Self>> {
```
[the remark this answers is not in the document]
Comment (lucab) [reply src/weekly/mod.rs:114]: That was in my initial implementation, but it was a bit too naïf thus failing on corner-cases.

In particular, a weekly calendar is a [finite field of order 7](https://en.wikipedia.org/wiki/Finite_field). Perhaps not very intuitive, but all the logic here has to obey [modular arithmetic](https://en.wikipedia.org/wiki/Modular_arithmetic) in that field. (same goes for hours and minutes too, really)

In other words, it has to account for the fact that a timespan can start on Sunday night and end on Monday morning, without jumping back in time :)

Splitting into multiple daily windows here makes the rest of the intervals logic easier to deal with, as it can simply ignore all modular details and just use integer arithmetic when comparing `MinuteInWeek`.

[the remark this answers is not in the document]
Comment (jlebon) [reply src/weekly/mod.rs:114]: Right, that makes sense. I guess we could split just the intervals which cross week boundaries into two. But meh... I see the appeal in being consistent.

[the remark this answers is not in the document]
Comment (lucab) [reply src/weekly/mod.rs:114]: Uh, good point, I somehow assumed it was easier to handle interval windows on daily boundaries but that's not a given. I'll try to quickly sketch your alternative next and check if it is easier to follow.

File: src/strategy/weekly.rs (hunk 2 above)
Comment (lucab) [reply src/strategy/weekly.rs:36]: For reference, https://github.com/main--/rust-intervaltree/issues/7 is in 0.2.6. I've left a TODO to compute the proper length measure here later on.

File: src/weekly/mod.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (lucab) [reply src/weekly/mod.rs:118]: I reworked the logic here to use week boundaries, and the final result is indeed easier to read. I've stick to that and pushed this new version, thanks for the suggestion!

**Comment by lucab** [comment]
@jlebon rebased with your suggestion, PTAL. I'm leaving the "compute length with overlaps" as a TODO as I wanted to re-check the algorithm for that first.

File: src/config/inputs.rs (hunk 1 above)
Comment (arithx) [reply src/config/inputs.rs:181]: `interval`?

Comment (dustymabe) [reply src/config/inputs.rs:181]: what was wrong with `periodic`? One way to handle adding another one later on is to use a version maybe

Comment (lucab) [reply src/config/inputs.rs:181]: @dustymabe nothing per se, it was just feeling more generic/broad than `weekly`. But if the latter is deceiving and the former seems ok, I can revert back.

(no need to worry for future strategy names here, I just wanted to note that this may not be the only time-based one)

File: docs/usage/updates-strategy.md [hunk 5]
```diff
@@ -56,3 +56,40 @@ Such an approach is only recommended where nodes are already grouped into an orc
...
+length_minutes = 30
+```
+
+The above configuration would result in three maintenance windows during which Zincati is allowed to reboot the machine for updates:
+ * 60 minutes starting at 23:30 UTC on Saturday night, and ending at 00:30 UTC on Sunday morning
+ * 60 minutes starting at 23:30 UTC on Sunday night, and ending at 00:30 UTC on Monday morning
+ * 30 minutes starting at 01:00 UTC on Wednesday morning, and ending at 01:30 UTC on Wednesday morning
+
+Reboot windows can be separately configured in multiple snippets, as long as each `updates.weekly.window` entry contains all the required properties:
+ * `days`: an array of weekdays (C locale), either in full or abbreviated form
```
Comment (zonggen) [inline docs/usage/updates-strategy.md:91]: super minor: could mention abbreviated = first three letters of full form

File: src/config/inputs.rs (hunk 1 above)
Comment (jlebon) [reply src/config/inputs.rs:181]: Will throw another entry in the hat: `weekly_schedule`. It's more wordy, but it makes it more clear that it's about setting up a schedule tied to a week rather than a single window which happens every week.

File: src/strategy/weekly.rs [hunk 6]
```diff
@@ -0,0 +1,122 @@
+//! Strategy for weekly weekly updates.
```
Comment (jlebon) [inline src/strategy/weekly.rs:1]: repeated word

File: src/weekly/utils.rs [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+        .saturating_add(u32::from(minute.min(59)));
+
+    assert!(weekly_minute < MAX_WEEKLY_MINS);
+    weekly_minute
+}
+
+/// Check duration for a sane lower and upper bound (whole week).
+pub(crate) fn check_duration(length: Duration) -> Fallible<Duration> {
+    // Whole week duration, in seconds.
+    const MAX_WEEKLY_SECS: u64 = MAX_WEEKLY_MINS as u64 * 60;
```
Comment (jlebon) [inline src/weekly/utils.rs:39]: This is already defined in the `weekly` module.

File: src/weekly/utils.rs [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+        "fri" | "friday" => Weekday::Fri,
+        "sat" | "saturday" => Weekday::Sat,
+        "sun" | "sunday" => Weekday::Sun,
+        _ => bail!("unrecognized week day: {}", input),
+    };
+
+    Ok(day)
+}
+
+/// Parse a time string (in 24h format).
```
Comment (jlebon) [inline src/weekly/utils.rs:67]: Would help to have a sample time string here.

File: src/strategy/weekly.rs [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+use log::trace;
+use serde::Serialize;
+use std::pin::Pin;
+use std::time::Duration;
+
+/// Strategy for weekly weekly updates.
+#[derive(Clone, Debug, Default, Serialize)]
+pub(crate) struct StrategyWeekly {
+    /// Whitelisted time windows during which updates are allowed.
+    allowed_intervals: WeeklyCalendar,
```
Comment (jlebon) [inline src/strategy/weekly.rs:17]: How about calling this e.g. `schedule`? It feels funny to name it something that refers to the underlying implementation when the struct name is meant to abstract it.

File: src/weekly/mod.rs [hunk 10]
```diff
@@ -0,0 +1,411 @@
...
+    pub fn containing_windows(&self, datetime: &DateTime<Utc>) -> Vec<&WeeklyWindow> {
+        let timepoint = utils::datetime_as_weekly_minute(datetime);
+        self.windows
+            .query_point(timepoint)
+            .map(|elem| &elem.value)
+            .collect()
+    }
+}
+
+impl Serialize for WeeklyCalendar {
```
Comment (jlebon) [inline src/weekly/mod.rs:106]: Oh interesting. Why do we need a serialization support for this?

File: src/weekly/mod.rs [hunk 11]
```diff
@@ -0,0 +1,411 @@
...
+    ) -> Fallible<Vec<Self>> {
+        failure::ensure!(
+            start_hour <= 24 && start_minute <= 59,
+            "invalid start time: {}:{}",
+            start_hour,
+            start_minute
+        );
+
+        let mut windows = vec![];
+        let mut remaining_length = utils::check_duration(length)?;
```
Comment (jlebon) [inline src/weekly/mod.rs:155]: Hmm, this makes it look like it's transforming its input. I'd personally just make the input parameter a `mut` and just call `utils::check_duration(length)?;` on its own.

File: src/weekly/mod.rs [hunk 12]
```diff
@@ -0,0 +1,411 @@
...
+        let mut start_weekly_minute =
+            utils::time_as_weekly_minute(start_day, start_hour, start_minute);
+        while remaining_length.as_secs() > 0 {
```
Comment (jlebon) [inline src/weekly/mod.rs:158]: This is an interesting construction. Really this can iterate at most twice, right? It might be clearer to just change `chop_timestan` to return a Vec of either one or two windows.

Comment (lucab) [reply src/weekly/mod.rs:158]: Good catch! Both the while-loop and most of `chop_timespan` logic are not actually useful anymore after moving away from daily windows. I moved some code around and dropped both.

File: src/weekly/mod.rs (hunk 11 above)
Comment (lucab) [reply src/weekly/mod.rs:155]: Fair point, it's a bit deceiving. I changed the helper signature to immutably borrow the argument, as it doesn't really need ownership there.

File: src/weekly/mod.rs (hunk 10 above)
Comment (lucab) [reply src/weekly/mod.rs:106]: We do embed this in the final runtime configuration, whose top-level container has a `Serialize` annotation.

I don't think it's used right now. I used to have some private logic for runtime config introspection, but I didn't turn into a public API as I wasn't unsure about the details. I can try resurrecting that if it seems useful.
Keeping the datastructure serializable is a low incremental effort and unlocks such future usage.

File: src/weekly/utils.rs (hunk 8 above)
Comment (lucab) [reply src/weekly/utils.rs:64]: Compiler checked examples in docstrings FTW!

File: src/config/inputs.rs (hunk 1 above)
Comment (lucab) [reply src/config/inputs.rs]: Ok, it looks like we have full circle without a clear winner. It seems that much confusion came from my late decision to step away from the initial pick. I have reverted everything back to use `periodic` as in the original form. How does it look like?

**Comment by lucab** [comment]
Rebased again, with a revert to the original `periodic` name, PTAL (and thanks all for the feedback so far).

**Comment by lucab** [comment]
Thanks everybody for the reviews/feedbacks! I'm holding this for a last a round of time-based manual testing, will merge afterwards unless anything is found to be majestically broken.

**Comment by lucab** [comment]
Manually tested on top of a `31.20200420.2.0` -> `31.20200505.2.0` update on `testing`, all good. Merging.
